### ISSUE TITLE
Reduce allocations from FieldCache

### ISSUE THREAD

**Issue by theolivenbaum** [apache/lucenenet#370]
Was doing today some more benchmark on our usage of Lucene, and noticed the current design of FieldCache allocates an anonymous class [here](https://github.com/apache/lucenenet/blob/37d7ffdede8918f3bce728647f438de47bd229be/src/Lucene.Net/Search/FieldCacheImpl.cs#L1347):

`````csharp
NumericDocValues valuesIn = reader.GetNumericDocValues(field);
if (valuesIn != null)
{
    // Not cached here by FieldCacheImpl (cached instead
    // per-thread by SegmentReader):
    return new FieldCache.Int64s(get: (docID) => valuesIn.Get(docID));   // <--- allocates due to capturing valuesIn
}
`````

In our usage of Lucene, we end up hitting this quite often, as you can see on the trace:
![image](https://user-images.githubusercontent.com/8791811/96337735-5f473c00-1089-11eb-880d-1aa849ff7ce6.png)

I'll investigate if there is any quick win here, and open a PR if so.

**Comment by theolivenbaum** [comment]
@NightOwl888 pushed the PR yesterday, #371, let me know if ok or should change something

**Comment by NightOwl888** [comment]
Fixed in #389
